### PR TITLE
mock device property injection for test

### PR DESCRIPTION
**Pull request by billypchan**
### Issues

Currently, to test the UI updating when device orientation or size class changes, we have to try it manually on simulator or real device. When we want to test different behavior on different idioms, we have to rerun the same tests to cover the idiom logic related code on different simulators. 

### Causes

For example, in LandingViewController we get device idiom from UIDevice.current.userInterfaceIdiom and current size class from self.traitCollection.horizontalSizeClass. Mocking size class change events and userInterfaceIdiom is not possible.

### Solutions

Create UIIdiomSizeClassOrientationProtocol and injects UIIdiomSizeClassOrientation as a param in the init method. For the unit test, injects MockIdiomSizeClassOrientation instead. Then we can mock the devices properties update events and write tests to check UI updates as we want.

## Notes

This PR is part of https://github.com/wireapp/wire-ios/pull/1640. A new test for NetworkStatusViewController size class changing will be written.

### CODE REVIEW

**Comment by vytis** [comment]
I think there is a super easy way to override trait collections by calling `setOverrideTraitCollection(_:forChildViewController:)`. So basically you just need to override the `userInterfaceIdiom`? 

The approach here on injecting dependencies is very convoluted I would say. The easiest way to do it is to create a protocol that exposes the method you need and use that instead of concrete type. There are plenty of examples in other places how we do it, just one that I found quickly:
https://github.com/wireapp/wire-ios/blob/26591d081c987d678ba0cf555a273bd96b15aa8e/Wire-iOS/Sources/UserInterface/Conversation/InputBar/CameraKeyboard/PhotoPermissionsController.swift#L28

**Comment by billypchan** [comment]
@vytis I get your point, I will refactor the code to match the pattern.

File: Wire-iOS/Sources/UserInterface/Registration/RegistrationTeam/LandingViewController.swift [hunk 1]
```diff
@@ -32,6 +32,8 @@ final class LandingViewController: UIViewController {
 
     private let tracker = AnalyticsTracker(context: AnalyticsContextRegistrationEmail)
 
+    fileprivate var uIIdiomSizeClassOrientationProtocol: UIIdiomSizeClassOrientationProtocol.Type
```
Comment (daehn) [inline Wire-iOS/Sources/UserInterface/Registration/RegistrationTeam/LandingViewController.swift]: I would suggest either capitalising both letter at the beginning or none.

File: Wire-iOS/Sources/Components/UserInterfaceIdiom.swift [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+protocol UserInterfaceIdiomProtocol {
+    static var idiom: UIUserInterfaceIdiom {get}
```
Comment (vytis) [inline Wire-iOS/Sources/Components/UserInterfaceIdiom.swift]: In this case the simplest approach is to create a protocol that `UIDevice` would conform without any changes. This way you can pass `UIDevice` where `UserInterfaceIdiomProtocol` is expected or your mock implementation that returns the `userInterfaceIdiom` you want.

File: Wire-iOS/Sources/UserInterface/Registration/RegistrationTeam/LandingViewController.swift [hunk 3]
```diff
@@ -144,6 +146,19 @@ final class LandingViewController: UIViewController {
         return button
     }()
 
+    /// init method for injecting mock UserInterfaceIdiomProtocol
+    ///
+    /// - Parameter userInterfaceIdiom: a struct Type compliant with UserInterfaceIdiomProtocol. Default Type is UserInterfaceIdiom.self. Provide this param for testing only.
+    init(userInterfaceIdiom: UserInterfaceIdiomProtocol.Type = UserInterfaceIdiom.self) {
```
Comment (vytis) [inline Wire-iOS/Sources/UserInterface/Registration/RegistrationTeam/LandingViewController.swift]: In here you could the just do `UserInterfaceIdiomProtocol = UIDevice.current`

File: Wire-iOS Tests/LandingViewControllerTests.swift [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    var mockUserInterfaceIdiom: MockUserInterfaceIdiom! = MockUserInterfaceIdiom()
+
+    override func setUp() {
+        super.setUp()
+        sut = LandingViewController(idiom: mockUserInterfaceIdiom)
+        mockParentViewControler.addChildViewController(sut)
+    }
+    
+    override func tearDown() {
+        super.tearDown()
```
Comment (vytis) [inline Wire-iOS Tests/LandingViewControllerTests.swift]: tearDown should be called at the end of the method

File: Wire-iOS/Sources/UserInterface/Registration/RegistrationTeam/LandingViewController.swift [hunk 5]
```diff
@@ -144,6 +146,19 @@ final class LandingViewController: UIViewController {
         return button
     }()
 
+    /// init method for injecting mock UserInterfaceIdiomProtocol
+    ///
+    /// - Parameter userInterfaceIdiom: a struct Type compliant with UserInterfaceIdiomProtocol. Default Type is UserInterfaceIdiom.self. Provide this param for testing only.
```
Comment (vytis) [inline Wire-iOS/Sources/UserInterface/Registration/RegistrationTeam/LandingViewController.swift]: The comment is outdated and I think a bit verbose - it's probably enough to just say that it's used for testing

**Pull request merged**